File: ambeth/Ambeth.Util/ambeth/typeinfo/EmbeddedRelationInfoItem.cs
using De.Osthus.Ambeth.Annotation;
using System;
using System.Text;

namespace De.Osthus.Ambeth.Typeinfo
{
    public class EmbeddedRelationInfoItem : IEmbeddedTypeInfoItem, IRelationInfoItem
    {
        protected IRelationInfoItem childMember;

        protected ITypeInfoItem[] memberPath;

        protected String name;

        public EmbeddedRelationInfoItem(String name, IRelationInfoItem childMember, params ITypeInfoItem[] memberPath)
        {
            this.name = name;
            this.childMember = childMember;
            this.memberPath = memberPath;
        }

        public IRelationInfoItem ChildMember
        {
            get
            {
                return childMember;
            }
        }

        ITypeInfoItem IEmbeddedTypeInfoItem.ChildMember
        {
            get
            {
                return childMember;
            }
        }

        public ITypeInfoItem[] MemberPath
        {
            get
            {
                return memberPath;
            }
        }

        public String MemberPathString
        {
            get
            {
                StringBuilder sb = new StringBuilder();
                foreach (ITypeInfoItem member in MemberPath)
                {
                    if (sb.Length > 0)
                    {
                        sb.Append('.');
                    }
                    sb.Append(member.Name);
                }
                return sb.ToString();
            }
        }

        public String[] MemberPathToken
        {
            get
            {
                String[] token = new String[MemberPath.Length];
                for (int a = MemberPath.Length; a-- > 0; )
                {
                    ITypeInfoItem member = MemberPath[a];
                    token[a] = member.Name;
                }
                return token;
            }
        }

        public Type DeclaringType
        {
            get
            {
                return childMember.DeclaringType;
            }
        }

        public Object DefaultValue
        {
            get
            {
                return childMember.DefaultValue;
            }
            set
            {
                childMember.DefaultValue = value;
            }
        }

        public Object NullEquivalentValue
        {
            get
            {
                return childMember.NullEquivalentValue;
            }
            set
            {
                childMember.NullEquivalentValue = value;
            }
        }

        public Type RealType
        {
            get
            {
                return childMember.RealType;
            }
        }

        public Type ElementType
        {
            get
            {
                return childMember.ElementType;
            }
        }

        public bool TechnicalMember
        {
            get
            {
                return false;
            }
            set
            {
                throw new NotSupportedException("A relation can never be a technical member");
            }
        }

        public Object GetValue(Object obj)
        {
            return GetValue(obj, false);
        }

        public Object GetValue(Object obj, bool allowNullEquivalentValue)
        {
            Object currentObj = obj;
            for (int a = 0, size = memberPath.Length; a < size; a++)
            {
                ITypeInfoItem memberPathItem = memberPath[a];
                currentObj = memberPathItem.GetValue(currentObj, allowNullEquivalentValue);
                if (currentObj == null)
                {
                    if (allowNullEquivalentValue)
                    {
                        return childMember.NullEquivalentValue;
                    }
                    return null;
                }
            }
            return childMember.GetValue(currentObj, allowNullEquivalentValue);
        }

        public void SetValue(Object obj, Object value)
        {
            Object currentObj = obj;
            for (int a = 0, size = memberPath.Length; a < size; a++)
            {
                ITypeInfoItem memberPathItem = memberPath[a];
                Object childObj = memberPathItem.GetValue(currentObj, false);
                if (childObj == null)
                {
                    childObj = Activator.CreateInstance(memberPathItem.RealType);
                    memberPathItem.SetValue(currentObj, childObj);
                }
                currentObj = childObj;
            }
            childMember.SetValue(currentObj, value);
        }

        public V GetAnnotation<V>() where V : Attribute
        {
            return childMember.GetAnnotation<V>();
        }

        public bool CanRead
        {
            get
            {
                return childMember.CanRead;
            }
        }

        public bool CanWrite
        {
            get
            {
                return childMember.CanWrite;
            }
        }

        public String Name
        {
            get
            {
                return name;
            }
        }

        public String XMLName
        {
            get
            {
                return childMember.XMLName;
            }
        }

        public bool IsXMLIgnore
        {
            get
            {
                return childMember.IsXMLIgnore;
            }
        }

        public override String ToString()
        {
            return "Embedded: " + Name + "/" + XMLName + " " + childMember;
        }

        public CascadeLoadMode CascadeLoadMode
        {
            get
            {
                return childMember.CascadeLoadMode;
            }
        }

        public bool IsManyTo
        {
            get
            {
                return childMember.IsManyTo;
            }
        }

        public bool IsToMany
        {
            get
            {
                return childMember.IsToMany;
            }
        }
    }
}

File: jambeth/jAmbeth.Util/src/main/java/de/osthus/ambeth/typeinfo/EmbeddedRelationInfoItem.java
package de.osthus.ambeth.typeinfo;

import java.lang.annotation.Annotation;
import java.util.Collection;

import de.osthus.ambeth.annotation.CascadeLoadMode;
import de.osthus.ambeth.exception.RuntimeExceptionUtil;

public class EmbeddedRelationInfoItem implements IEmbeddedTypeInfoItem, IRelationInfoItem
{
	protected IRelationInfoItem childMember;

	protected ITypeInfoItem[] memberPath;

	protected String name;

	public EmbeddedRelationInfoItem(String name, IRelationInfoItem childMember, ITypeInfoItem... memberPath)
	{
		this.name = name;
		this.childMember = childMember;
		this.memberPath = memberPath;
	}

	@Override
	public IRelationInfoItem getChildMember()
	{
		return childMember;
	}

	@Override
	public ITypeInfoItem[] getMemberPath()
	{
		return memberPath;
	}

	@Override
	public String getMemberPathString()
	{
		StringBuilder sb = new StringBuilder();
		for (ITypeInfoItem member : getMemberPath())
		{
			if (sb.length() > 0)
			{
				sb.append('.');
			}
			sb.append(member.getName());
		}
		return sb.toString();
	}

	@Override
	public String[] getMemberPathToken()
	{
		ITypeInfoItem[] memberPath = getMemberPath();
		String[] token = new String[memberPath.length];
		for (int a = memberPath.length; a-- > 0;)
		{
			ITypeInfoItem member = memberPath[a];
			token[a] = member.getName();
		}
		return token;
	}

	@Override
	public Class<?> getDeclaringType()
	{
		return childMember.getDeclaringType();
	}

	@Override
	public Object getDefaultValue()
	{
		return childMember.getDefaultValue();
	}

	@Override
	public void setDefaultValue(Object defaultValue)
	{
		childMember.setDefaultValue(defaultValue);
	}

	@Override
	public Collection<?> createInstanceOfCollection()
	{
		return childMember.createInstanceOfCollection();
	}

	@Override
	public Object getNullEquivalentValue()
	{
		return childMember.getNullEquivalentValue();
	}

	@Override
	public void setNullEquivalentValue(Object nullEquivalentValue)
	{
		childMember.setNullEquivalentValue(nullEquivalentValue);
	}

	@Override
	public Class<?> getRealType()
	{
		return childMember.getRealType();
	}

	@Override
	public Class<?> getElementType()
	{
		return childMember.getElementType();
	}

	@Override
	public Object getValue(Object obj)
	{
		return getValue(obj, false);
	}

	@Override
	public Object getValue(Object obj, boolean allowNullEquivalentValue)
	{
		Object currentObj = obj;
		for (int a = 0, size = memberPath.length; a < size; a++)
		{
			ITypeInfoItem memberPathItem = memberPath[a];
			currentObj = memberPathItem.getValue(currentObj, allowNullEquivalentValue);
			if (currentObj == null)
			{
				if (allowNullEquivalentValue)
				{
					return childMember.getNullEquivalentValue();
				}
				return null;
			}
		}
		return childMember.getValue(currentObj, allowNullEquivalentValue);
	}

	@Override
	public void setValue(Object obj, Object value)
	{
		Object currentObj = obj;
		for (int a = 0, size = memberPath.length; a < size; a++)
		{
			ITypeInfoItem memberPathItem = memberPath[a];
			Object childObj = memberPathItem.getValue(currentObj, false);
			if (childObj == null)
			{
				try
				{
					childObj = memberPathItem.getRealType().newInstance();
				}
				catch (InstantiationException e)
				{
					throw RuntimeExceptionUtil.mask(e);
				}
				catch (IllegalAccessException e)
				{
					throw RuntimeExceptionUtil.mask(e);
				}
				memberPathItem.setValue(currentObj, childObj);
			}
			currentObj = childObj;
		}
		childMember.setValue(currentObj, value);
	}

	@Override
	public <V extends Annotation> V getAnnotation(Class<V> annotationType)
	{
		return childMember.getAnnotation(annotationType);
	}

	@Override
	public boolean canRead()
	{
		return childMember.canRead();
	}

	@Override
	public boolean canWrite()
	{
		return childMember.canWrite();
	}

	@Override
	public boolean isTechnicalMember()
	{
		return false;
	}

	@Override
	public void setTechnicalMember(boolean technicalMember)
	{
		throw new UnsupportedOperationException("A relation can never be a technical member");
	}

	@Override
	public String getName()
	{
		return name;
	}

	@Override
	public String getXMLName()
	{
		return childMember.getXMLName();
	}

	@Override
	public boolean isXMLIgnore()
	{
		return childMember.isXMLIgnore();
	}

	@Override
	public String toString()
	{
		return "Embedded: " + getName() + "/" + getXMLName() + " " + childMember;
	}

	@Override
	public CascadeLoadMode getCascadeLoadMode()
	{
		return childMember.getCascadeLoadMode();
	}

	@Override
	public boolean isManyTo()
	{
		return childMember.isManyTo();
	}

	@Override
	public boolean isToMany()
	{
		return childMember.isToMany();
	}
}
